python/lsst/daf/butler/_formatter.py

```diff
@@ -51,12 +63,1034 @@
+
+        Raises
+        ------
+        RuntimeError
+            Raised if validation fails. The default implementation raises
+            if any recipes are given.
+        """
+        if recipes:
+            raise RuntimeError(f"This formatter does not understand these write recipes: {recipes}")
+        return recipes
+
+    @classmethod
+    def name(cls) -> str:
+        """Return the fully qualified name of the formatter.
+
+        Returns
+        -------
+        name : `str`
+            Fully-qualified name of formatter class.
+        """
+        return get_full_type_name(cls)
+
+    def _is_disassembled(self) -> bool:
+        """Return `True` if this formatter is looking at a disassembled
+        component.
+        """
+        return self.file_descriptor.component is not None
+
+    def _check_resource_size(self, uri: ResourcePath, recorded_size: int, resource_size: int) -> None:
+        """Compare the recorded size with the resource size.
+
+        The given URI will not be accessed.
+        """
+        if recorded_size >= 0 and resource_size != recorded_size:
+            raise FileIntegrityError(
+                "Integrity failure in Datastore. "
+                f"Size of file {uri} ({resource_size}) "
+                f"does not match size recorded in registry of {recorded_size}"
+            )
+
+    def _get_cache_ref(self) -> DatasetRef:
+        """Get the `DatasetRef` to use for cache look ups.
+
+        Returns
+        -------
+        ref : `lsst.daf.butler.DatasetRef`
+            The dataset ref to use when looking in the cache.
+            For single-file dataset this will be the dataset ref directly.
+            If this is disassembled we need the component and the component
+             will be in the `FileDescriptor`.
+        """
+        if self.file_descriptor.component is None:
+            cache_ref = self.dataset_ref
+        else:
+            cache_ref = self.dataset_ref.makeComponentRef(self.file_descriptor.component)
+        return cache_ref
+
+    def _ensure_cache(
+        self, cache_manager: AbstractDatastoreCacheManager | None = None
+    ) -> AbstractDatastoreCacheManager:
+        """Return the cache if given else return a null cache."""
+        if cache_manager is None:
+            # Circular import avoidance.
+            from .datastore.cache_manager import DatastoreDisabledCacheManager
+
+            cache_manager = DatastoreDisabledCacheManager(None, None)
+        return cache_manager
+
+    def read(
+        self,
+        component: str | None = None,
+        expected_size: int = -1,
+        cache_manager: AbstractDatastoreCacheManager | None = None,
+    ) -> Any:
+        """Read a Dataset.
+
+        Parameters
+        ----------
+        component : `str`, optional
+            Component to read from the file. Only used if the `StorageClass`
+            for reading differed from the `StorageClass` used to write the
+            file.
```

Comment:
The read vs write storage class definitions haven't moved. They are still in `FileDescriptor`.  The write storage class is the storage class that was used to write the dataset -- it could be the component storage class or the composite depending on whether it was disassembled. You can tell if it was disassembled because `FileDescriptor.component` will be defined.

The read storage class used to only differ from the write one for disassembly but now with storage class conversion they can be different intentionally. I need to make sure that Eli is getting the different value so that if someone stored pandas data frame they can get back a different thing without first reading it as pandas.

Reply:
I added some extra documentation on the storage class overrides and how they appear to the formatter. I've verified that if you ask for an override it will always turn up in the readStorageClass.